File: Java/src/javaProj/Game.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package javaProj;

import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;
import java.util.stream.Collectors;

/**
 *
 * @author The user
 */
public class Game {

    Scanner in = new Scanner(System.in);

    private String playerName;
    private static int totalScore = 0;
    private static int colorScore = 0;
    private static int suitScore = 0;
    private static int numberScore = 0;

    public Game(String playerName) {
        this.playerName = playerName;
    }

    //A function that manages the game
    public void startGame(int deckSize) {
        System.out.println("Welcome To The Guess Game...");
        System.out.println(this.playerName + " , Please follow the instructions.");
        Deck deck = new Deck(deckSize);
        for (int i = 0; i < deckSize; i++) {
            Card card = deck.getCard();
            this.turn(card.getColor(), card.getNumber(), card.getSuit());
            System.out.println("The real card is:");
            card.drawCard();
            if (i % 2 == 1) {
                deck.fetchRandomSentence();
            }
        }
        System.out.println("Champion!! You finished the game. ");
        System.out.println("________________________");
        System.out.println("Color Score: " + this.colorScore +"/"+ deckSize);
        System.out.println("Suit Score: " + this.suitScore +"/"+ deckSize);
        System.out.println("Number Score: " + this.numberScore +"/"+ deckSize);
        System.out.println("________________________");
        System.out.println("Total Score:" + this.totalScore +"/"+ deckSize * 3);
        System.out.println("________________________");
    }

    //For each turn, it receives the necessary inputs, compares them to the real card, and responds accordingly.
    public void turn(String Ccolor, int Cnumber, String Csuit) {
        System.out.print("Guess the color: ");
        String color = in.nextLine().toLowerCase();

        System.out.print("Guess the suit: ");
        String suit = in.nextLine().toLowerCase();

        System.out.print("Guess the number: ");
        int number = in.nextInt();
        in.nextLine();// To clear the remaining character in the buffer

        Map<String, Boolean> results = new HashMap<>();// Create a results map
        results.put("color", Ccolor.equalsIgnoreCase(color));
        results.put("suit", Csuit.equalsIgnoreCase(suit));
        results.put("number", Cnumber == number);

        long correctCount = results.values().stream().filter(Boolean::booleanValue).count();// Calculate the number of matches

        if (correctCount == 3) {
            System.out.println("You beat it!! All the guesses are correct");
            totalScore += 3;
            colorScore++;
            System.out.println("color++");
            suitScore++;
            System.out.println("suit++");
            numberScore++;
            System.out.println("number++");
        } else if (correctCount == 2) {
            String correctItems = results.entrySet().stream()
                    .filter(Map.Entry::getValue)
                    .map(Map.Entry::getKey)
                    .collect(Collectors.joining(" and "));
            System.out.println("You almost beat it!! " + correctItems + " are correct");
            totalScore += 2;
            if (results.get("color")) {
                colorScore++;
                System.out.println("color++");
            }
            if (results.get("suit")) {
                suitScore++;
                System.out.println("suit++");
            }
            if (results.get("number")) {
                numberScore++;
                            System.out.println("number++");

            }
        } else if (correctCount == 1) {
            String correctItem = results.entrySet().stream()
                    .filter(Map.Entry::getValue)
                    .map(Map.Entry::getKey)
                    .findFirst().orElse("");
            System.out.println("You almost beat it!! " + correctItem + " is correct");
            totalScore += 1;
            if (results.get("color")) {
                colorScore++;
                System.out.println("color++");
            }
            if (results.get("suit")) {
                suitScore++;
                System.out.println("suit++");

            }
            if (results.get("number")) {
                numberScore++;
                            System.out.println("number++");

            }
        } else {
            System.out.println("What a loser, everything was wrong.");
        }
    }

}

File: C++/Game.cpp
#include "Game.h"
#include <iostream>
#include <string>
#include <algorithm>
#include <map>
using namespace std;

Game::Game(string playerName):playerName(playerName),totalScore(0),colorScore(0),suitScore(0),numberScore(0)
{}

//A function that manages the game
void Game::startGame(int deckSize)
{
    cout<<"Welcome To The Guess Game...\n";
    cout << this->playerName << " , Please follow the instructions.\n";
    Deck deck =Deck(deckSize);
    for (int i = 0; i < deckSize; i++) {
        Card card = deck.getCard();
        turn(card.getColor(), card.getNumber(), card.getSuit());
        cout << "The real card is:\n";
        card.drawCard();
        if (i % 2 == 1) {
            deck.fetchRandomSentence();
        }
    }
    cout << "Champion!! You finished the game. \n";
    cout << ("________________________\n");
    cout << "Color Score: " << this->colorScore << "/" << deckSize<<endl;
    cout << "Suit Score: " << this->suitScore <<"/" << deckSize<<endl;
    cout << "Number Score: " << this->numberScore << "/" << deckSize<<endl;
    cout << "________________________\n";
    cout << "Total Score:" <<this->totalScore <<"/" << deckSize * 3<<endl;
    cout << "________________________\n";
}

//For each turn, it receives the necessary inputs, compares them to the real card, and responds accordingly.
void Game::turn(const string& Ccolor, int Cnumber, const string& Csuit) {
    string color, suit;
    int number;
    cin.ignore();  // To clear the remaining character in the buffer
    cout << "Guess the color: ";
    getline(cin, color);
    transform(color.begin(), color.end(), color.begin(), ::tolower);

    cout << "Guess the suit: ";
    getline(cin, suit);
    transform(suit.begin(), suit.end(), suit.begin(), ::tolower);

    cout << "Guess the number: ";
    cin >> number;
    cin.ignore();  // To clear the remaining character in the buffer

    map<string, bool> results;// Create a results map
    results["color"] = (Ccolor == color);
    results["suit"] = (Csuit == suit);
    results["number"] = (Cnumber == number);

    int correctCount = count_if(results.begin(), results.end(),
        [](const auto& pair) { return pair.second; });// Calculate the number of matches

    // Output results based on correct guesses
    if (correctCount == 3) {
        cout << "You beat it!! All the guesses are correct\n";
        totalScore += 3;
        colorScore++;
        suitScore++;
        numberScore++;
    }
    else if (correctCount == 2) {
        string correctItems;
        for (auto it = results.begin(); it != results.end(); ++it) {
            if (it->second) {
                if (!correctItems.empty()) correctItems += " and ";
                correctItems += it->first;
            }
        }
       cout << "You almost beat it!! " << correctItems << " are correct\n";
        totalScore += 2;
        if (results["color"]) {
            colorScore++;
        }
        if (results["suit"]) {
            suitScore++;
        }
        if (results["number"]) {
            numberScore++;
        }
    }
    else if (correctCount == 1) {
        for (auto it = results.begin(); it != results.end(); ++it) {
            if (it->second) {
                cout << "You almost beat it!! " << it->first << " is correct\n";
                break;
            }
        }
            totalScore += 1;
            if (results["color"]) {
                colorScore++;
            }
            if (results["suit"]) {
                suitScore++;
            }
            if (results["number"]) {
                numberScore++;
            }
        }
    else {
        cout << "What a loser, everything was wrong.\n";
    }
    }
